### ISSUE TITLE
how to get  for   each audio  frame's   log-probability  on  every  element of   the  alphabet   table ?

### ISSUE THREAD

**Issue by bolt163** [mozilla/DeepSpeech#1390]
in the  deepspeech.cc's  infer function there is a  float type 3-dimensional array  **input_data_mat** , in   which:

the  1st   dimension equals  aNFrames(frames's mfcc of the current audio wav) )
the  2nd  dimension 1
the  3rd   dimension :  num_classes, equals  count of element in the alphabet.txt table; 

values of the array  which were  also  feed to the decode  function of  deepspeech.cc , I have a doubt of the releated of input_data_mat array,   it was not storing  each audio  frame's  log-probability on  every  element of   the  alphabet  table ? 
I dump the  the array's value into file  and find there were lots of  minus value, (for each frame  the dimension is correct),  so whats' the mathmatical  mean of the   input_data_mat's value

![log-prob](https://user-images.githubusercontent.com/25170435/40544873-4962cfc4-605c-11e8-95a5-f8ee50a55649.png)

Do  there  need any transformation to get log-probability from  input_data_mat array? thanks

https://github.com/mozilla/DeepSpeech/blob/master/native_client/deepspeech.cc#L337

------------------------------------------original code-----------------------------
DEEPSPEECH_EXPORT
char*
Model::infer(float* aMfcc, int aNFrames, int aFrameLen)
{
  const int batch_size = BATCH_SIZE;
  const int timesteps = aNFrames;
  const size_t num_classes = mPriv->alphabet->GetSize() + 1; // +1 for blank

  const int frameSize = mPriv->ncep + (2 * mPriv->ncep * mPriv->ncontext);

  float*** input_data_mat = (float***)calloc(timesteps, sizeof(float**));
  for (int i = 0; i < timesteps; ++i) {
    input_data_mat[i] = (float**)calloc(batch_size, sizeof(float*));
    for (int j = 0; j < batch_size; ++j) {
      input_data_mat[i][j] = (float*)calloc(num_classes, sizeof(float));
    }
  }

  if (mPriv->run_aot) {
#ifdef DS_NATIVE_MODEL
    Eigen::ThreadPool tp(2);  // Size the thread pool as appropriate.
    Eigen::ThreadPoolDevice device(&tp, tp.NumThreads());

    nativeModel nm(nativeModel::AllocMode::RESULTS_PROFILES_AND_TEMPS_ONLY);
    nm.set_thread_pool(&device);

    for (int ot = 0; ot < timesteps; ot += DS_MODEL_TIMESTEPS) {
      nm.set_arg0_data(&(aMfcc[ot * frameSize]));
      nm.Run();

      // The CTCDecoder works with log-probs.
      for (int t = 0; t < DS_MODEL_TIMESTEPS, (ot + t) < timesteps; ++t) {
        for (int b = 0; b < batch_size; ++b) {
          for (int c = 0; c < num_classes; ++c) {
            input_data_mat[ot + t][b][c] = nm.result0(t, b, c);
          }
        }
      }
    }
#else
    std::cerr << "No support for native model built-in." << std::endl;
    return NULL;
#endif // DS_NATIVE_MODEL
  } else {
    if (aFrameLen == 0) {
      aFrameLen = frameSize;
    } else if (aFrameLen < frameSize) {
      std::cerr << "mfcc features array is too small (expected " <<
        frameSize << ", got " << aFrameLen << ")\n";
      return NULL;
    }

    Tensor input(DT_FLOAT, TensorShape({1, aNFrames, frameSize}));

    auto input_mapped = input.tensor<float, 3>();
    for (int i = 0, idx = 0; i < aNFrames; i++) {
      for (int j = 0; j < frameSize; j++, idx++) {
        input_mapped(0, i, j) = aMfcc[idx];
      }
      idx += (aFrameLen - frameSize);
    }

    Tensor n_frames(DT_INT32, TensorShape({1}));
    n_frames.scalar<int>()() = aNFrames;

    // The CTC Beam Search decoder takes logits as input, we can feed those from
    // the "logits" node in official models or
    // the "logits_output_node" in old AOT hacking models
    std::vector<Tensor> outputs;
    Status status = mPriv->session->Run(
      {{ "input_node", input }, { "input_lengths", n_frames }},
      {"logits"}, {}, &outputs);

    // If "logits" doesn't exist, this is an older graph. Try to recover.
    if (status.code() == tensorflow::error::NOT_FOUND) {
      status.IgnoreError();
      status = mPriv->session->Run(
        {{ "input_node", input }, { "input_lengths", n_frames }},
        {"logits_output_node"}, {}, &outputs);
    }

    if (!status.ok()) {
      std::cerr << "Error running session: " << status.ToString() << "\n";
      return NULL;
    }

    auto logits_mapped = outputs[0].tensor<float, 3>();
    // The CTCDecoder works with log-probs.
    for (int t = 0; t < timesteps; ++t) {
      for (int b = 0; b < batch_size; ++b) {
        for (int c = 0; c < num_classes; ++c) {
          input_data_mat[t][b][c] = logits_mapped(t, b, c);
        }
      }
    }
  }

  return decode(aNFrames, input_data_mat);
}

**Comment by bolt163** [comment]
has solved by myself
